dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/GitHubSourceCodeRepo.java

```diff
@@ -315,69 +287,150 @@ public Workflow setupWorkflowVersions(String repositoryId, Workflow workflow, Op
-                            } else if (identifiedType.equals(SourceFile.FileType.NEXTFLOW_CONFIG)) {
-                                testJson.setType(SourceFile.FileType.NEXTFLOW_TEST_PARAMS);
-                            }
-
-                            testJson.setPath(workflow.getDefaultTestParameterFilePath());
-                            testJson.setAbsolutePath(workflow.getDefaultTestParameterFilePath());
-                            testJson.setContent(testJsonContent);
-
-                            // Check if test parameter file has already been added
-                            long duplicateCount = version.getSourceFiles().stream().filter((SourceFile v) -> v.getPath().equals(workflow.getDefaultTestParameterFilePath()) && v.getType() == testJson.getType()).count();
-                            if (duplicateCount == 0) {
-                                version.getSourceFiles().add(testJson);
-                            }
-                        }
-                    }
+                sha = ref.getObject().getSha();
+                if (ref.getObject().getType().equals("tag")) {
+                    GHTagObject tagObject = repository.getTagObject(sha);
+                    sha = tagObject.getObject().getSha();
+                } else if (ref.getObject().getType().equals("branch")) {
+                    GHBranch branch = repository.getBranch(refName);
+                    sha = branch.getSHA1();
                 }
 
-            } catch (Exception ex) {
-                LOG.info(gitUsername + ": " + workflow.getDefaultWorkflowPath() + " on " + ref + " was not valid workflow", ex);
+                GHCommit commit = repository.getCommit(sha);
+                branchDate = commit.getCommitDate();
+                if (branchDate.before(epochStart)) {
+                    branchDate = epochStart;
+                }
+            } catch (IOException e) {
+                LOG.info("unable to retrieve commit date for branch " + refName);
             }
+            return Triple.of(refName, branchDate, sha);
+        } else {
+            return null;
+        }
+    }
 
-            version = versionValidation(version, workflow, calculatedPath);
 
-            workflow.addWorkflowVersion(version);
-        }
+    /**
+     * Creates a workflow version for a specific version within a workflow
+     * @param repositoryId Unique repository ID (ex. dockstore/dockstore-ui2)
+     * @param workflow Workflow object
+     * @param ref Triple containing reference name, branch date, and SHA
+     * @param existingWorkflow Optional existing workflow
+     * @param existingDefaults Optional mapping of existing versions
+     * @param repository GitHub repository object
+     * @return WorkflowVersion for the given reference
+     */
+    private WorkflowVersion setupWorkflowVersionsHelper(String repositoryId, Workflow workflow, Triple<String, Date, String> ref, Optional<Workflow> existingWorkflow,
```

Comment:
Ya I didn't like using the triple, but again I don't want to refactor existing code that much since it is likely that the refresh is going to see more refactoring in the near future. I simply separated out reusable code for this new endpoint.

I do agree that we should use a class here or do it in a cleaner way.

Reply:
I agree. Triple and Pair and useful, but my rule of thumb is that they should be internal to methods or at least parameters for private methods only. 
Once they get exposed to any external methods, they should be eliminated